src/interp.rs

```diff
@@ -19,153 +19,483 @@
-                .collect(),
-            z: z.into_boxed_slice(),
-            zi: 0,
+        pub fn process(&mut self, src: &[f32], dst: &mut [f32]) {
+            assert!(src.len().checked_mul(self.factor) == Some(dst.len()));
+            assert!(self.z.len() == self.delay * self.channels as usize);
+            assert!(self.filter.len() == self.factor);
+            assert!(self.zi < self.delay);
+
+            if src.is_empty() {
+                return;
+            }
+
+            let frames = src.len() / self.channels as usize;
+
+            for (src, (dst, z)) in src.chunks_exact(frames).zip(
+                dst.chunks_exact_mut(frames * self.factor)
+                    .zip(self.z.chunks_exact_mut(self.delay)),
+            ) {
+                let mut zi = self.zi;
+
+                for (src, dst) in src.iter().zip(dst.chunks_exact_mut(self.factor)) {
+                    // Add sample to delay buffer
+                    //
+                    // TODO Ringbuffer without bounds checks for z/zi
+                    //
+                    // Safety: zi is checked to be between 0 and self.delay
+                    *unsafe { z.get_unchecked_mut(zi) } = *src;
+
+                    // Apply coefficients
+                    for (filter, dst) in self.filter.iter().zip(dst.iter_mut()) {
+                        let mut acc = 0.0;
+                        for (c, index) in &*filter.coeff {
+                            let mut i = zi as i32 - *index as i32;
+                            if i < 0 {
+                                i += self.delay as i32;
+                            }
+
+                            // Safety: zi is checked to be between 0 and self.delay
+                            acc += *unsafe { z.get_unchecked(i as usize) } as f64 * c;
+                        }
+
+                        *dst = acc as f32;
+                    }
+
+                    zi += 1;
+                    if zi == self.delay {
+                        zi = 0;
+                    }
+                }
+            }
+
+            self.zi = (self.zi + frames) % self.delay;
         }
     }
+}
 
-    pub fn get_factor(&self) -> usize {
-        self.factor
+/// A trait to be generic over number of channels in a of frame
+trait Frame: Sized + Copy {
+    const CHANNELS: usize;
+
+    fn scale_add(&mut self, other: &Self, coeff: f32);
+    fn from_planar(slice: &[f32], stride: usize) -> Self;
+}
+
+type MonoFrame32 = [f32; 1];
+type StereoFrame32 = [f32; 2];
+type SurroundFrame32 = [f32; 6];
+
+impl Frame for MonoFrame32 {
+    const CHANNELS: usize = 1;
+
+    #[inline(always)]
+    fn scale_add(&mut self, other: &Self, coeff: f32) {
+        self[0] += other[0] * coeff;
     }
 
-    pub fn reset(&mut self) {
-        // TODO: Use slice::fill() once stabilized
-        for v in &mut *self.z {
-            *v = 0.0;
-        }
-        self.zi = 0;
+    #[inline(always)]
+    fn from_planar(slice: &[f32], _stride: usize) -> Self {
```

Comment:
I wonder if for the specialized versions it would actually be faster to not deinterleave the input first. Did you check that?

Reply:
Slightly faster, probably. (It benefits from interleaved data, since it allows in-cpu parallel computation over the channels)

However, I think we could overlook that for now. Part 2 in the optimizations will be about changing how the true_peak calculations use the interpolator. Short story is that there is no need to deal with more than one input-sample at a time to the interpolator. It already has its internal 12/24-frame buffer, and all we want is to figure out the max generated sample. Avoiding an output buffer both reduces allocations and improves memory-locality.

Reply:
Sounds good to me :)